core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -430,6 +430,41 @@ public ShareFetchContext newContext(String groupId, Map<TopicIdPartition, ShareF
         return context;
     }
 
+    /**
+     * The acknowledgeSessionUpdate method is used to update the request epoch and lastUsed time of the share session.
+     * @param groupId The group id in the share fetch request.
+     * @param reqMetadata The metadata in the share acknowledge request.
+     */
+    public void acknowledgeSessionUpdate(String groupId, ShareFetchMetadata reqMetadata) {
+        if (reqMetadata.epoch() == ShareFetchMetadata.INITIAL_EPOCH) {
+            // ShareAcknowledge Request cannot have epoch as INITIAL_EPOCH (0)
+            throw Errors.INVALID_SHARE_SESSION_EPOCH.exception();
+        } else if (reqMetadata.epoch() == ShareFetchMetadata.FINAL_EPOCH) {
+            ShareSessionKey key = shareSessionKey(groupId, reqMetadata.memberId());
+            if (cache.remove(key) != null) {
+                log.debug("Removed share session with key " + key);
+            }
```

Comment:
Thanks for the review. I think you are right. We should check for the existence of the share sessions in the final epoch, not just in case of shareAcknowledge, but also in case of shareFetch. In case of shareFetch we might have piggybacked acknowledgements in the final epoch and if the session has already been closed / evicted, the current code would allow the acknowledgements to proceed. I will make the required changes both in `SharePartitionManager.newContext` and `SharePartitionManager.acknowledgeSessionUpdate`. I would also request @AndrewJSchofield and @apoorvmittal10 to also validate my understanding, thanks !

Reply:
Could we simplify the above code as the following?

```
            if (cache.remove(key) == null) {
                log.error("Share session error for {}: no such share session found", key);
                throw Errors.SHARE_SESSION_NOT_FOUND.exception();
            } else {
                log.debug("Removed share session with key " + key);
            }
```